packages/firestore/src/util/bundle_reader.ts

```diff
@@ -0,0 +1,229 @@
+   *
+   * Once a complete element is read, it is dropped from internal buffer.
+   *
+   * Returns either the bundled element, or null if we have reached the end of the stream.
+   */
+  private async readNextElement(): Promise<SizedBundleElement | null> {
+    const lengthBuffer = await this.readLength();
+    if (lengthBuffer === null) {
+      return null;
+    }
+
+    const lengthString = this.textDecoder.decode(lengthBuffer);
+    const length = Number(lengthString);
+    if (isNaN(length)) {
+      this.raiseError(`length string (${lengthString}) is not valid number`);
+    }
+
+    const jsonString = await this.readJsonString(lengthBuffer.length, length);
+    // Update the internal buffer to drop the read length and json string.
+    this.buffer = this.buffer.slice(lengthBuffer.length + length);
+
+    return new SizedBundleElement(
+      JSON.parse(jsonString),
+      lengthBuffer.length + length
+    );
+  }
+
+  // First index of '{' from the underlying buffer.
+  private indexOfOpenBracket(): number {
+    return this.buffer.findIndex(v => v === '{'.charCodeAt(0));
+  }
+
+  // Reads from the beginning of the internal buffer, until the first '{', and return
+  // the content.
+  // If reached end of the stream, returns a null.
+  private async readLength(): Promise<Uint8Array | null> {
+    let position = this.indexOfOpenBracket();
+    while (position < 0) {
+      const bytesRead = await this.pullMoreDataToBuffer();
+      if (bytesRead < 0) {
+        if (this.buffer.length === 0) {
+          return null;
+        }
+        position = this.indexOfOpenBracket();
+        // Underlying stream is closed, and we still cannot find a '{'.
+        if (position < 0) {
+          this.raiseError(
+            'Reached the end of bundle when a length string is expected.'
+          );
+        }
+      } else {
+        position = this.indexOfOpenBracket();
+      }
+    }
+
+    return this.buffer.slice(0, position);
+  }
+
+  // Reads from a specified position from the internal buffer, for a specified
+  // number of bytes, pulling more data from the underlying stream if needed.
+  //
+  // Returns a string decoded from the read bytes.
+  private async readJsonString(start: number, length: number): Promise<string> {
+    while (this.buffer.length < start + length) {
+      const bytesRead = await this.pullMoreDataToBuffer();
+      if (bytesRead < 0) {
+        this.raiseError('Reached the end of bundle when more is expected.');
+      }
+    }
+
+    return this.textDecoder.decode(this.buffer.slice(start, start + length));
+  }
+
+  private raiseError(message: string): void {
+    // eslint-disable-next-line @typescript-eslint/no-floating-promises
+    this.reader.cancel('Invalid bundle format.');
+    throw new Error(message);
+  }
+
+  // Pulls more data from underlying stream to internal buffer.
+  // Returns a boolean indicating whether the stream is finished.
```

Comment:
If you keep this as is, please update the method comment to stay that it returns the length read. I did notice however that this information is not used, so we could keep the boolean.